_posts/2021-08-31-modifier-type-clef-primaire-non-referencee-vers-biginit.md

```diff
@@ -0,0 +1,257 @@
+
+Une fois la migration achevée, une fenêtre de maintenance est requise (1h dans le [cas d'origine](https://tech.coffeemeetsbagel.com/schema-migration-from-int-to-bigint-on-a-massive-table-in-postgres-aabb835c3b84)) pour :
+
+- supprimer les contraintes sur la colonne INTEGER (PK, FK) ;
+- promouvoir la colonne de type BIGINT en tant qu'identifiant ;
+- recréer les contraintes (PK et FK).
+
+### Améliorations
+
+Cette solution va dans le bon sens, mais ne satisfait par tous nos critères : il reste une fenêtre de maintenance. Nous pensons qu'il est possible de pousser la solution et d'éliminer totalement cette fenêtre.
+
+Le principe est le suivant :
+
+- la contrainte de clef primaire comporte une contrainte NOT NULL et une contrainte UNIQUE ;
+- c'est la validation de ces contraintes qui prend le plus de temps dans la fenêtre de maintenance ;
+- ces validations peuvent être effectuées en amont.
+
+Pour valider les contraintes en amont :
+
+- ajouter une contrainte NOT NULL dès la création de la colonne ;
+- créer un index UNIQUE de manière concurrente [`CREATE INDEX CONCURRENTLY`](https://www.postgresql.org/docs/current/sql-createindex.html).
+
+Il ne reste plus, comme dernière opération, qu'à créer la PK sur base de l'index UNIQUE.
+`ALTER TABLE "foo" ADD CONSTRAINT "foo_pkey" PRIMARY KEY USING INDEX "foo_id_unique"`
+
+PostgreSQL détecte lui-même la contrainte NOT NULL.
+
+## Implémentation
+
+Le code source étant trop verbeux pour figurer intégralement dans cet article, la solution est présentée en pseudo SQL/NodeJS sur une table imaginaire, avec un lien vers le commit de notre repository pour les détails.
+
+A l'état initial, la table contient 700 millions d'enregistrements.
+[fd7273f9](https://github.com/1024pix/pix/commit/fd7273f9e47648e4615a8b4af13bd81b1c8c2928#diff-754627687e8d1bd3c4d8f60cf5ed3640a9c104acdccbc20407c16479ad63055d)
+
+``` sql
+CREATE TABLE foo (ID INTEGER PRIMARY KEY);
+INSERT INTO foo generate_series(1, 700 000 000);
+```
+
+La première étape est l'ajout d'une colonne temporaire, avec une migration de BDD, inclue dans le déploiement de la release.
+[eecfd941](https://github.com/1024pix/pix/pull/3357/files#diff-1eecfd94150cca766655d3786c7dc12e3b6bc8fa60be3a0c0740424ab02c157a)
+
+```sql
+ALTER TABLE foo ADD COLUMN "idBigInteger" NOT NULL DEFAULT -1;
+CREATE TRIGGER WHEN INSERT INTO foo FOR EACH ROW :NEW."idBigInteger" = :new.id;
+```
+
+Ensuite, le script de migration inclus dans la PR est lancé manuellement (durée: 48h pour 700 millions de lignes).
+[d89cb206](https://github.com/1024pix/pix/pull/3357/files#diff-d89cb20637cb9b836bce4902b46db958e0209c2a0de96e7dc81a0c43015680b4)
+
+```javascript
+const maxid = await client.query('SELECT MAX(id) INTO maxId FROM foo');
+for (const i = 0; i <= maxId; i++){
+  await client.query(`UPDATE foo SET "idBigInteger" = id WHERE id BETWEEN i AND (i + ${chunkSize} -1)`);
+}
+await client.query('CREATE UNIQUE INDEX CONCURRENTLY "indexidBigInteger"');
+```
```

Comment:
```suggestion
const chunkSize = 10000000;
const {rows: [{maxId}]} = await client.query('SELECT MAX(id) "maxId" FROM foo');
for (let startId = 0; startId <= maxId; startId += chunkSize) {
  await client.query(`UPDATE foo SET "idBigInteger" = id WHERE id BETWEEN ${startId} AND ${startId + chunkSize - 1}`);
}
await client.query('CREATE UNIQUE INDEX CONCURRENTLY "idx_bigint" ON foo("idBigInteger")');
```

Même opinion sur le pseudo-JS que sur le pseudo-SQL. Je ne pense pas que glisser des erreurs comme `SELECT MAX(id) INTO` ou `for(const i` aide à la lecture. On peut proposer du code simplifié qui reste correct.

Reply:
Yep, voir https://github.com/1024pix/pix-engineering/pull/19#discussion_r701637398